### PR TITLE
Return plugin error when requesting settings

### LINKED ISSUE #85519: Improve plugin state management
One big weak point of managing plugins in the backend is our lack of ability to understand what state the plugin is in. We don't have a concept for this and our only distinction between a plugin that we tried to load couldn't and one that we could is whether it is part of the Plugin Manager's map.

There are various different states that would be useful to track as it can help us choose more appropriate paths/directions for a plugin to take.

The goal of this issue is to be able to associate a state to a plugin, so in case of error, a more user-friendly error is returned to the user. For example: `Plugin unavailable: Missing signature`.

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

Before this change, a failed plugin could still load or get just the error "Plugin unavailable":

[Screencast from 12-04-24 15:05:38.webm](https://github.com/grafana/grafana/assets/4025665/118217bd-2471-464d-8011-1534974cd098)

After this PR, the `/plugins/:pluginId/settings` endpoint checks if there is an error associated to the plugin and, in that case, it returns the public message associated with that error:

image

**Which issue(s) does this PR fix?**:

Closes https://github.com/grafana/grafana/issues/85519

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/pluginsintegration/pipeline/pipeline.go [hunk 1]
```diff
@@ -65,10 +65,10 @@ func ProvideInitializationStage(cfg *config.PluginManagementCfg, pr registry.Ser
 		InitializeFuncs: []initialization.InitializeFunc{
 			ExternalServiceRegistrationStep(cfg, externalServiceRegistry, tracer),
 			initialization.BackendClientInitStep(pluginEnvProvider, bp),
-			initialization.PluginRegistrationStep(pr),
 			initialization.BackendProcessStartStep(pm),
 			RegisterPluginRolesStep(roleRegistry),
 			ReportBuildMetrics,
+			initialization.PluginRegistrationStep(pr),
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/pipeline/pipeline.go:71]: this changes ensure that the registration step is done the last one so only healthy plugins are registered

File: pkg/plugins/ifaces.go [hunk 2]
```diff
@@ -110,6 +110,7 @@ type StaticRouteResolver interface {
 
 type ErrorResolver interface {
```
Comment (wbrowne) [inline pkg/plugins/ifaces.go:111]: I haven't dug in further, but do you think we need to keep `ErrorResolver` long term when we have `ErrorTracker`?

**Review by wbrowne — COMMENTED** [review]
Just a few questions!

File: pkg/plugins/models.go [hunk 3]
```diff
@@ -296,6 +296,23 @@ func (e *Error) WithMessage(m string) *Error {
 	return e
 }
 
+func (e Error) PublicMessage() string {
```
Comment (sympatheticmoose) [inline pkg/plugins/models.go:299]: I just want to check - these are all fatal error states which would leave the plugin unable to successfully run, including if the configuration allows for unsigned plugins etc?

Comment (sympatheticmoose) [reply pkg/plugins/models.go:299]: perfect - thanks!

File: pkg/services/pluginsintegration/pluginerrs/errors.go [hunk 4]
```diff
@@ -28,6 +28,16 @@ func (s *Store) PluginErrors(ctx context.Context) []*plugins.Error {
 	return errs
 }
 
+func (s *Store) PluginError(ctx context.Context, pluginID string) *plugins.Error {
+	err := s.errs.Error(ctx, pluginID)
+	if err == nil {
+		return nil
+	}
+
+	err.ErrorCode = err.AsErrorCode()
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/services/pluginsintegration/pluginerrs/errors.go:37]: Yeah let's figure it out later then once we have something in place!

**Review by wbrowne — APPROVED** [review]

**Pull request merged**